src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CommandMessageBinaryEncoder.cs

```diff
@@ -296,7 +296,10 @@ private void WriteType1Section(BsonBinaryWriter writer, Type1CommandMessageSecti
             stream.WriteCString(section.Identifier);
 
             var batch = section.Documents;
-            var maxDocumentSize = section.MaxDocumentSize ?? writer.Settings.MaxDocumentSize;
+            var maxDocumentSize = 
+                IsEncryptionConfigured && MaxDocumentSize.HasValue
+                    ? MaxDocumentSize.Value 
+                    : section.MaxDocumentSize ?? writer.Settings.MaxDocumentSize;
```

Comment:
Is this the proper solution? Feels like just another variation of hard coding.

Seems like the issue is that there are two ways that `MaxDocumentSize` gets communicated. In the past they have always agreed so we never noticed that there were two ways.

How hard would it be to ensure that there is only one way to configure the `MaxDocumentSize`?

Reply:
The problem here, that this value is filled before the time when we modify `MaxDocumentSize` and `maxMessageSize`, so currently another option which I see is to have two places when we modify `MessageEncoderSettings`. 
One is here: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/48/files#diff-fd9dc09b71b78534cd6783fd7f6f8eeaR177
And another one is somewhere here: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Operations/RetryableInsertCommandOperation.cs#L139
or in the line before this method.

And probably the same steps for non-retryable operations..

I don't think that this solution better than the currently implemented.

Reply:
This quick and dirty fix might be the best short term solution, even if it isn't "right".

Rather than fix the underlying issue we can go with this workaround, which is essentially just saying that if encryption is configured use the `MaxDocumentSize` from the `MessageEncoderSettings` instead of the `section.MaxDocumentSize`.

It does nothing to resolve other cases where the two values might disagree.

Reply:
Is there ever a case where `IsEncryptionConfigured == true && MaxDocumentSize.HasValue==false`?

Reply:
https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Operations/RetryableInsertCommandOperation.cs#L139
If `channel.ConnectionDescription.MaxDocumentSize` is null here. I don't know whether it's possible or not.